[TensorExpr] Make Load and Store multi-dimensional.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35800 [TensorExpr] Make Load and Store multi-dimensional.**

This PR includes the following changes:
* Introduce a new `Expr` type `Buf`: it plays a similar to `Var` role, but also has dimensions.
* Use the new `Buf` class in `Store` and `Load` instead of `Var` for specifying where to store to or load from. `Buf` contains the dimensions info of the buffer we're loading/storing to and hence we are able to keep N-d indexes without flattening them into a 1-d index ([x,y] vs [x+y*W]).
* Flattening of the indexes is now a separate pass that is executed in `LoopNest::prepareForCodegen` - backends still expect indexes to be flattened, and this PR preserves that.
* `Tensor` now contains a `Buf` instead of `Var`, and thus Tensor now has the dimensions info (previously it was a property of a `Function`, not a `Tensor`). This brings us closer to Tensor being a combination of Buffer + Function, where Buffer specifies iteration domain and the Function defines a computation.

TODOs:
* Consider merging `Buffer` with `Buf` or `BufHandle`. It seems that we don't need all of them.
* Harden the logic of how we create buffers in fuser pass. Currently it seems that sometimes we don't set dimensions.
* Use `Buf` in `Allocate` and `Free`.
* Make it clearer that `Function` doesn't "own" dimensions info and that dimensions are a property of a Tensor, not a Function.

Differential Revision: [D20789005](https://our.internmc.facebook.com/intern/diff/D20789005)

Commit:
[WIP][TensorExpr] Make Load and Store multi-dimensional.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Make Load and Store multi-dimensional."


This PR includes the following changes:
* Introduce a new `Expr` type `Buf`: it plays a similar to `Var` role, but also has dimensions.
* Use the new `Buf` class in `Store` and `Load` instead of `Var` for specifying where to store to or load from. `Buf` contains the dimensions info of the buffer we're loading/storing to and hence we are able to keep N-d indexes without flattening them into a 1-d index ([x,y] vs [x+y*W]).
* Flattening of the indexes is now a separate pass that is executed in `LoopNest::prepareForCodegen` - backends still expect indexes to be flattened, and this PR preserves that.
* `Tensor` now contains a `Buf` instead of `Var`, and thus Tensor now has the dimensions info (previously it was a property of a `Function`, not a `Tensor`). This brings us closer to Tensor being a combination of Buffer + Function, where Buffer specifies iteration domain and the Function defines a computation.

TODOs:
* Consider merging `Buffer` with `Buf` or `BufHandle`. It seems that we don't need all of them.
* Harden the logic of how we create buffers in fuser pass. Currently it seems that sometimes we don't set dimensions.
* Use `Buf` in `Allocate` and `Free`.
* Make it clearer that `Function` doesn't "own" dimensions info and that dimensions are a property of a Tensor, not a Function.

Differential Revision: [D20789005](https://our.internmc.facebook.com/intern/diff/D20789005)

[ghstack-poisoned]

zheng-xq commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -818,7 +818,7 @@ void LLVMCodeGenImpl::visit(const Load* v) {
     v->base_handle()->accept(this);
     auto base = this->value_;
-    v->index()->accept(this);
+    v->flattened_index()->accept(this);
```
I think the index-flattening should be part of the common-mode lowering, before sending to individual backends. So the backends do not have to deal with this complexity. In other words, 

* Schedule transformation: 
  * All the indices are multi-dimensional tensor index
* When lower is called: 
  * First, common-mode lower
    * Convert the tensor indices into flattened index
    * Apply transformations such as vectorization
  * Send to device-specific codegen, where the index is flattened, and vectorized.

ZolotukhinM replied:
It works now approximately as you described, this call here does not perform the flattening - what it does is it tries to access the flattened index. If the indexes have not been lowered at that point, this call will error out.

zheng-xq commented on test/cpp/tensorexpr/test_cuda.cpp:
```diff
@@ -161,8 +161,8 @@ void testCudaDynamicShape2D() {
     VarHandle n("n", kInt);
-    Buffer a(VarHandle("a", kHandle), kFloat, {m, n});
-    Buffer b(VarHandle("b", kHandle), kFloat, {m, n});
+    Buffer a(BufHandle("a", kHandle, {m, n}), kFloat);
```
BufHandle takes a kHandle is a bit weird. It is somewhat expected for the original VarHandle, since it is a much lower level API. Some unfamiliar might expect BufHandle would imply its own handle value, and the Dtype is that of the value, or the handle itself.

ZolotukhinM replied:
Right, I noticed that part too and it seemed redundant, but I wasn't sure it was. I can remove the `kHandle` argument I think.

zheng-xq commented on torch/csrc/jit/tensorexpr/ir.cpp:
```diff
@@ -10,46 +10,190 @@ static Dtype ChooseDtype(const Dtype& buffer_dtype, const Dtype& index_dtype) {
+      continue;
+    }
+    ExprHandle stride = ExprHandle(dims[i + 1]);
+    for (size_t j = i + 2; j < ndim; j++) {
```
Maybe comment on why not just something like? 

stride[i] = stride[i + 1] * dims[i + 1];

ZolotukhinM replied:
I've just revisited this part - hopefully the new version is simpler!

zheng-xq commented on torch/csrc/jit/tensorexpr/ir.cpp:
```diff
@@ -10,46 +10,190 @@ static Dtype ChooseDtype(const Dtype& buffer_dtype, const Dtype& index_dtype) {
+    strides[i] = stride;
+  }
+
+  ExprHandle total_index = int32_t{0};
```
Why "int32_t{0}"?

ZolotukhinM replied:
The code for computing `total_index` was there before, I copied it almost verbatim :)

Commit:
Update on "[TensorExpr] Make Load and Store multi-dimensional."


This PR includes the following changes:
* Introduce a new `Expr` type `Buf`: it plays a similar to `Var` role, but also has dimensions.
* Use the new `Buf` class in `Store` and `Load` instead of `Var` for specifying where to store to or load from. `Buf` contains the dimensions info of the buffer we're loading/storing to and hence we are able to keep N-d indexes without flattening them into a 1-d index ([x,y] vs [x+y*W]).
* Flattening of the indexes is now a separate pass that is executed in `LoopNest::prepareForCodegen` - backends still expect indexes to be flattened, and this PR preserves that.
* `Tensor` now contains a `Buf` instead of `Var`, and thus Tensor now has the dimensions info (previously it was a property of a `Function`, not a `Tensor`). This brings us closer to Tensor being a combination of Buffer + Function, where Buffer specifies iteration domain and the Function defines a computation.

TODOs:
* Consider merging `Buffer` with `Buf` or `BufHandle`. It seems that we don't need all of them.
* Harden the logic of how we create buffers in fuser pass. Currently it seems that sometimes we don't set dimensions.
* Use `Buf` in `Allocate` and `Free`.
* Make it clearer that `Function` doesn't "own" dimensions info and that dimensions are a property of a Tensor, not a Function.

Differential Revision: [D20789005](https://our.internmc.facebook.com/intern/diff/D20789005)

[ghstack-poisoned]

Commit:
clang-format on "[TensorExpr] Make Load and Store multi-dimensional."


This PR includes the following changes:
* Introduce a new `Expr` type `Buf`: it plays a similar to `Var` role, but also has dimensions.
* Use the new `Buf` class in `Store` and `Load` instead of `Var` for specifying where to store to or load from. `Buf` contains the dimensions info of the buffer we're loading/storing to and hence we are able to keep N-d indexes without flattening them into a 1-d index ([x,y] vs [x+y*W]).
* Flattening of the indexes is now a separate pass that is executed in `LoopNest::prepareForCodegen` - backends still expect indexes to be flattened, and this PR preserves that.
* `Tensor` now contains a `Buf` instead of `Var`, and thus Tensor now has the dimensions info (previously it was a property of a `Function`, not a `Tensor`). This brings us closer to Tensor being a combination of Buffer + Function, where Buffer specifies iteration domain and the Function defines a computation.

TODOs:
* Consider merging `Buffer` with `Buf` or `BufHandle`. It seems that we don't need all of them.
* Harden the logic of how we create buffers in fuser pass. Currently it seems that sometimes we don't set dimensions.
* Use `Buf` in `Allocate` and `Free`.
* Make it clearer that `Function` doesn't "own" dimensions info and that dimensions are a property of a Tensor, not a Function.

Differential Revision: [D20789005](https://our.internmc.facebook.com/intern/diff/D20789005)

[ghstack-poisoned]

zheng-xq commented on torch/csrc/jit/tensorexpr/expr.h:
```diff
@@ -158,6 +158,66 @@ class Var : public ExprNode<Var> {
+ public:
+  BufHandle() : ExprHandle(nullptr) {}
+  //   explicit BufHandle(Dtype dtype) : ExprHandle(Buf::make(dtype)) {}
+  BufHandle(const std::string& name_hint, const std::vector<ExprHandle>& dims)
```
Should BufHandle carry the dtype to its content?

ZolotukhinM replied:
Yes, eventually it should. In this PR I only made `Buf` := `Var` + `dims` (and dtype of the var is `kHandle`), but in follow-ups I plan to add dtype and then we might be able to merge it with `Buffer`.

zheng-xq commented on test/cpp/tensorexpr/test_expr.cpp:
```diff
@@ -69,9 +69,9 @@ void testExprLetStmtTest01() {
   Buffer b_buf("b", kFloat, {1});
 
-  ExprHandle load_a = Load::make(a_buf, 0, 1);
+  ExprHandle load_a = Load::make(a_buf, {0}, 1);
```
Minor: since this was very common, it might be easier to add an overloaded version that takes a single Index, instead of vector of indices.

ZolotukhinM replied:
Here I prefer being explicit (it might also help people to "relearn" how `Load` now works) - it was one of the most frequent changes in this PR, as I had to change all the existing code, but I don't expect it to be that much of annoyance going forward in incremental changes.

zheng-xq commented on torch/csrc/jit/tensorexpr/ir.cpp:
```diff
@@ -10,46 +10,183 @@ static Dtype ChooseDtype(const Dtype& buffer_dtype, const Dtype& index_dtype) {
+      ExprHandle(1).node());
+}
+
+const Expr* flattened_index(
```
Minor: I'd prefer flat_index for shorter name, instead of flattened_index. But it's up to you.

ZolotukhinM replied:
Sounds good, I'll rename!

Commit:
Update on "[TensorExpr] Make Load and Store multi-dimensional."


This PR includes the following changes:
* Introduce a new `Expr` type `Buf`: it plays a similar to `Var` role, but also has dimensions.
* Use the new `Buf` class in `Store` and `Load` instead of `Var` for specifying where to store to or load from. `Buf` contains the dimensions info of the buffer we're loading/storing to and hence we are able to keep N-d indexes without flattening them into a 1-d index ([x,y] vs [x+y*W]).
* Flattening of the indexes is now a separate pass that is executed in `LoopNest::prepareForCodegen` - backends still expect indexes to be flattened, and this PR preserves that.
* `Tensor` now contains a `Buf` instead of `Var`, and thus Tensor now has the dimensions info (previously it was a property of a `Function`, not a `Tensor`). This brings us closer to Tensor being a combination of Buffer + Function, where Buffer specifies iteration domain and the Function defines a computation.

TODOs:
* Consider merging `Buffer` with `Buf` or `BufHandle`. It seems that we don't need all of them.
* Harden the logic of how we create buffers in fuser pass. Currently it seems that sometimes we don't set dimensions.
* Use `Buf` in `Allocate` and `Free`.
* Make it clearer that `Function` doesn't "own" dimensions info and that dimensions are a property of a Tensor, not a Function.

Differential Revision: [D20789005](https://our.internmc.facebook.com/intern/diff/D20789005)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Make Load and Store multi-dimensional."


This PR includes the following changes:
* Introduce a new `Expr` type `Buf`: it plays a similar to `Var` role, but also has dimensions.
* Use the new `Buf` class in `Store` and `Load` instead of `Var` for specifying where to store to or load from. `Buf` contains the dimensions info of the buffer we're loading/storing to and hence we are able to keep N-d indexes without flattening them into a 1-d index ([x,y] vs [x+y*W]).
* Flattening of the indexes is now a separate pass that is executed in `LoopNest::prepareForCodegen` - backends still expect indexes to be flattened, and this PR preserves that.
* `Tensor` now contains a `Buf` instead of `Var`, and thus Tensor now has the dimensions info (previously it was a property of a `Function`, not a `Tensor`). This brings us closer to Tensor being a combination of Buffer + Function, where Buffer specifies iteration domain and the Function defines a computation.

TODOs:
* Consider merging `Buffer` with `Buf` or `BufHandle`. It seems that we don't need all of them.
* Harden the logic of how we create buffers in fuser pass. Currently it seems that sometimes we don't set dimensions.
* Use `Buf` in `Allocate` and `Free`.
* Make it clearer that `Function` doesn't "own" dimensions info and that dimensions are a property of a Tensor, not a Function.

Differential Revision: [D20789005](https://our.internmc.facebook.com/intern/diff/D20789005)

[ghstack-poisoned]
